lib/tasks/one_off.rake

```diff
@@ -25,4 +25,66 @@ namespace :one_off do
       end
     end
   end
+
+  # This task tracks down the original "published" dates for NuGet Versions that
+  # are marked as "published_at=1900-01-01 00:00:00" in Libraries, and updates
+  # them with the original "published" date, and marks them 'status: "Deprecated"'.
+  #
+  # It processes lists-of-lists from NuGet's API:
+  #   * [CatalogRoot] Catalog API (https://learn.microsoft.com/en-us/nuget/api/catalog-resource)
+  #     * .items => [CatalogPage]
+  #       * .items => [PackageDetails] (each PackageDetails may contain the original "published")
+  #
+  # @param catalog_start_idx Integer the index of the catalog page to start with,
+  #                          to skip already-processed pages (default: 0)
+  # @param threads Integer the number of parallel threads to process version items (default: 8)
+  desc "backfill unknown nuget published_at dates"
+  task :backfill_nuget_published_at, %i[catalog_start_idx threads] => :environment do |_t, args|
+    require "parallel"
+
+    catalog_start_idx = args.catalog_start_idx.to_i
+    threads = args.threads || 8
+
+    print "Loading 'unlisted' NuGet Version records... "
+    unlisted_name_versions = Version
+      .joins(:project)
+      .where(projects: { platform: "NuGet" })
+      .where("published_at < ?", DateTime.new(1901, 1, 1))
+      .pluck("projects.name, versions.number")
+      .index_by(&:itself)
+    puts "found #{unlisted_name_versions.size} records."
+
+    print "Loading NuGet's catalog listing of events pages... "
+    catalog = PackageManager::ApiService.request_json_with_headers("https://api.nuget.org/v3/catalog0/index.json")
+    catalog_size = catalog["items"].size
+    puts "found #{catalog_size} pages."
+
+    catalog["items"][catalog_start_idx..].each.with_index do |catalog_item, idx|
+      puts "Processing NuGet catalog page #{idx}, out of #{catalog_size} pages..."
+      page = PackageManager::ApiService.request_json_with_headers(catalog_item["@id"])
+
+      Parallel.each(page["items"], in_threads: threads, progress: "Page #{idx} (out of #{page['items'].size} items)") do |item|
+        page_item = PackageManager::ApiService.request_json_with_headers(item["@id"])
+        name = page_item["id"]
+        version = page_item["version"]
+        published = page_item["published"]
+
+        if published.nil?
+          raise "'published' field not found on page #{catalog_item['@id']} in #{item['@id']}."
+        elsif unlisted_name_versions.key?([version, name]) && published !~ /1900-/
+          p = Project.find_by(platform: "NuGet", name: name)
+          v = p.versions.find_by(number: version)
+          if v.published_at.year == 1900
+            published_at = Time.parse(published)
+            v.update_columns(published_at: published, status: "Deprecated")
```

Comment:
Might these versions be undeprecated if they now have a published_at > 1900?

Reply:
the "Catalogs" list of a list of all events, so the event we're looking at here would be an older one before the version was unlisted/1900'ed. 

I'm not sure if NuGet lets you re-list an "unlisted" version, but if that happened then we'll see it in the next `PackageManager::NuGet.update()` for the  package. This task is basically just responsible for swapping the `published_at: '1900-01-01 00:00:00'" == DEPRECATED` equivalency with `status: "Deprecated" == DEPRECATED`.

but that is a great question, because I should probably handle the case of "re-listing" in the NuGet updater here: https://github.com/librariesio/libraries.io/blob/49a931d438f13799d673b37b677ad31ff3559a44/app/models/package_manager/nu_get.rb#L160-L166